### PR TITLE
Prime BSI support

### LINKED ISSUE #1: Prime BSI support

### PR DESCRIPTION
**Pull request by aliddell**
Provide support for both Prime BSI and BSI Express models. Takes inspiration both from the [Spinnaker](https://github.com/acquire-project/acquire-driver-spinnaker) and [DCAM](https://github.com/acquire-project/acquire-driver-hdcam) drivers.

Adds tests for both models.

~~Populates the .github/ directory with the usual complement of workflows.~~

### CODE REVIEW

**Comment by aliddell** [comment]
Be aware that triggering is not ready yet.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+env:
+  BUILD_TYPE: Release
+
+jobs:
+  windows-build:
+    name: "Build on Windows with PVCAM SDK"
+    runs-on:
+      - self-hosted
+      - pvcam
+      - Prime-BSI # for tests ## FIXME (aliddell): we won't have both on one machine. We need a way to do either.
```
Comment (andy-sweet) [inline .github/workflows/release.yml]: Both means Prime-BSI and Prime-BSI-Express cameras?

If so, looks like you separate this dependency before merging?

File: tests/prime-bsi-express-one-video-stream.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    const auto next = [](VideoFrame* cur) -> VideoFrame* {
+        return (VideoFrame*)(((uint8_t*)cur) + cur->bytes_of_frame);
+    };
+
+    const auto consumed_bytes = [](const VideoFrame* const cur,
+                                   const VideoFrame* const end) -> size_t {
+        return (uint8_t*)end - (uint8_t*)cur;
+    };
+
+    clock clock_{};
```
Comment (andy-sweet) [inline tests/prime-bsi-express-one-video-stream.cpp]: Given the definitions in platform.h, I think the type needs the explicit struct.

```suggestion
    struct clock clock_{};
```

Maybe some preprocessors/compilers can infer that, but my version of Apple clang (14.0.3) could not, so I was unable to build before doing this.

File: tests/prime-bsi-express-one-video-stream.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    static double time_limit_ms = 20000.0;
+    clock_init(&clock_);
+    clock_shift_ms(&clock_, time_limit_ms);
+    OK(acquire_start(runtime));
+
+    uint64_t nframes = 0;
+    while (nframes < props.video[0].max_frame_count) {
+        clock throttle{};
```
Comment (andy-sweet) [inline tests/prime-bsi-express-one-video-stream.cpp]:
```suggestion
        struct clock throttle{};
```

File: tests/prime-bsi-one-video-stream.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    const auto next = [](VideoFrame* cur) -> VideoFrame* {
+        return (VideoFrame*)(((uint8_t*)cur) + cur->bytes_of_frame);
+    };
+
+    const auto consumed_bytes = [](const VideoFrame* const cur,
+                                   const VideoFrame* const end) -> size_t {
+        return (uint8_t*)end - (uint8_t*)cur;
+    };
+
+    clock clock_{};
```
Comment (andy-sweet) [inline tests/prime-bsi-one-video-stream.cpp]:
```suggestion
    struct clock clock_{};
```

File: tests/prime-bsi-one-video-stream.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    static double time_limit_ms = 20000.0;
+    clock_init(&clock_);
+    clock_shift_ms(&clock_, time_limit_ms);
+    OK(acquire_start(runtime));
+
+    uint64_t nframes = 0;
+    while (nframes < props.video[0].max_frame_count) {
+        clock throttle{};
```
Comment (andy-sweet) [inline tests/prime-bsi-one-video-stream.cpp]:
```suggestion
        struct clock throttle{};
```

File: tests/prime-bsi-abort-while-waiting-for-trigger.cpp [hunk 6]
```diff
@@ -0,0 +1,5 @@
+int
+main()
+{
+    return 0;
```
Comment (andy-sweet) [inline tests/prime-bsi-abort-while-waiting-for-trigger.cpp]: Since triggering is not being handled and this is just empty, remove from this PR and add back when supported.

File: tests/prime-bsi-express-configure-triggering.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
+// Checks various trigger setting manipulation
```
Comment (andy-sweet) [inline tests/prime-bsi-express-configure-triggering.cpp]: Since triggering is not being handled, remove from this PR and add back when supported.

File: tests/prime-bsi-configure-properties.cpp [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    ASSERT_EQ(uint32_t, "%d", props->video[0].camera.settings.shape.x, 2048);
+    ASSERT_EQ(uint32_t, "%d", props->video[0].camera.settings.shape.y, 2048);
+
+    props->video[0].camera.settings.pixel_type = SampleType_u16;
+    OK(acquire_configure(runtime, props));
+    ASSERT_EQ(SampleType,
+              "%d",
+              props->video[0].camera.settings.pixel_type,
+              SampleType_u16);
+    //        //        props->video[0].camera.settings.pixel_type =
```
Comment (andy-sweet) [inline tests/prime-bsi-configure-properties.cpp]: Uncomment or remove?

File: tests/prime-bsi-express-configure-properties.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    ASSERT_EQ(uint32_t, "%d", props->video[0].camera.settings.shape.x, 2048);
+    ASSERT_EQ(uint32_t, "%d", props->video[0].camera.settings.shape.y, 2048);
+
+    props->video[0].camera.settings.pixel_type = SampleType_u16;
+    OK(acquire_configure(runtime, props));
+    ASSERT_EQ(SampleType,
+              "%d",
+              props->video[0].camera.settings.pixel_type,
+              SampleType_u16);
+    //        //        props->video[0].camera.settings.pixel_type =
```
Comment (andy-sweet) [inline tests/prime-bsi-express-configure-properties.cpp]: Uncomment or remove?

File: tests/prime-bsi-express-metadata.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    DEVOK(device_manager_select(dm,
+                                DeviceKind_Storage,
+                                SIZED("Trash") - 1,
+                                &props.video[0].storage.identifier));
+
+    // Some metadata is dependent on some properties.
+    // Reset the properties to something sensible.
+    //    props.video[0].camera.settings.binning = 1;
+    //    props.video[0].camera.settings.offset = { .x = 0, .y = 0 };
+    //    props.video[0].camera.settings.shape = { .x = 2048, .y = 2048 };
```
Comment (andy-sweet) [inline tests/prime-bsi-express-metadata.cpp]: Uncomment or remove?

File: src/pvcam.cpp [hunk 11]
```diff
@@ -0,0 +1,1279 @@
...
+check_pvcam_camera_id(uint64_t id)
+{
+    constexpr uint64_t limit = std::numeric_limits<int16>::max();
+    EXPECT(id <= limit, "Expected a value at most %llu. Got %llu.", limit, id);
+}
+
+void
+camera_properties_to_rgn_type(CameraProperties* properties, rgn_type& roi)
+{
+    auto binning = (uns16)properties->binning;
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Where does the `uns16` definition/alias come from? The PVCAM headers?

Nit: is it equivalent to `uint16_t`? If so, could we use that instead for consistency with other code (both here and elsewhere) and for extra clarity?

Comment (andy-sweet) [reply src/pvcam.cpp]: Looks like there's a few other types like this (e.g. `int16`), so same question applies to all.

File: src/pvcam.cpp [hunk 12]
```diff
@@ -0,0 +1,1279 @@
...
+        if (it != pixel_format_to_sample_type.end()) {
+            meta->supported_pixel_types |= (1U << it->second);
+        }
+    }
+}
+
+void
+PVCamCamera::query_triggering_capabilities(CameraPropertyMetadata* meta) const
+{
+    char model[MAX_PRODUCT_NAME_LEN];
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Optional: since triggering support is not part of this PR, I'd consider removing this implementation (and others) if it's not too much trouble.

File: src/pvcam.cpp [hunk 13]
```diff
@@ -0,0 +1,1279 @@
...
+    void stop();
+    void execute_trigger() const;
+    void get_frame(void* im, size_t* nbytes, ImageInfo* info);
+
+  private:
+    // Guards access to the PVCAM API.
+    std::shared_ptr<std::mutex> pvcam_api_mutex_;
+
+    // The PVCAM camera handle.
+    int16 hcam;
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Why doesn't this private member have a trailing underscore?

```suggestion
    int16 hcam_;
```

Updates elsewhere will be needed, so you probably don't want to actually suggestion this suggestion through github.

File: src/pvcam.cpp [hunk 14]
```diff
@@ -0,0 +1,1279 @@
...
+    meta->offset.y = {
+        .writable = 1,
+        .low = (float)roi_min.p1,
+        .high = (float)roi_max.p1,
+        .type = PropertyType_FixedPrecision,
+    };
+
+    meta->shape.x = {
+        .writable = 1,
+        .low = (float)(roi_min.s2 - roi_max.s1 + 1),
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Is it guaranteed that `roi_min.s2 >= roi_max.s1`?

I assume `roi_min.s2` is the upper bound on the left x-coordinate of the region and `roi_max.s1` is the lower bound on the right x-coordinate of the region, so I could imagine cases where that condition is not guaranteed (i.e. because i can define small regions that start on the right of the sensor, and small regions that end on the left of the sensor).

In that case, I think this calculation in general could produce surprising values (i.e. a large intermediate unsigned result). Though maybe some/most compilers will handle this more gracefully than I give them credit for?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Generally, looks great. I made a few suggestions, but most are optional or are nits.

The one thing I'd like to see (ideally before) merging is a run of the PR test action so I can see all the checks passing since I don't have a machine/camera to test this with.

File: src/pvcam.cpp (hunk 13 above)
Comment (aliddell) [reply src/pvcam.cpp]: Thanks!

File: src/pvcam.cpp (hunk 11 above)
Comment (aliddell) [reply src/pvcam.cpp]: Yeah, it comes from the PVCAM headers. I went with using the data types used by the PVCAM headers when I expect them to be used with / returned from PVCAM API calls. I can see how it would be confusing, and it's almost certain they won't change their typedefs, so I can change them to `u?int\d\d_t` types.

File: src/pvcam.cpp (hunk 14 above)
Comment (aliddell) [reply src/pvcam.cpp]: My understanding was that `roi_min.s2` is the lower bound on the right x-coordinate of the region and `roi_max.s1` is the upper bound on the left x-coordinate of the region. However, one thing I missed in the SDK docs about `ATTR_MIN` and `ATTR_MAX`, though:

> The value is only valid for the following data types: TYPE_INT8, TYPE_UNS8, TYPE_INT16, TYPE_UNS16, TYPE_INT32, TYPE_UNS32, TYPE_INT64, TYPE_UNS64, TYPE_FLT64 and TYPE_BOOLEAN.

This is sort of born out in the return values I get, which give me

```c++
roi_max = {
  .s1 = 0,
  .s2 = 2047,
  .sbin = 1,
  .p1 = 0,
  .p2 = 2047,
  .pbin = 1
};

roi_min = {
  .s1 = 0,
  .s2 = 0,
  .sbin = 1,
  .p1 = 0,
  .p2 = 0,
  .pbin = 1
};
```

Looking at this again, and I can only guess at this since the documentation isn't clear, I have to imagine that only `.s2` and `.p2` are set for the max value, and only `.s1` and `.p1` are set for the min value. `.sbin` and `.pbin` are queryable with a different parameter. So it doesn't make sense to use `roi_max.s1` or `roi_min.p2`, for example. We know that the highest values we can set for `.s2` and `.p2` are 2047 (the sensor width and height), and we'd like to know for sure that the highest value you can set for `.s1` or `.p1` is 2047, but that's not clear. I can set the width offset to that in PVCamTest with a width of 1, and it's fine. I don't know what restrictions there are on other models, though.

I took another look at the parameters and I don't believe there are any others to query. You can get serial and parallel sensor sizes, but these are fixed.

Comment (aliddell) [reply src/pvcam.cpp]: See updated comment ^

File: src/pvcam.cpp [hunk 15]
```diff
@@ -0,0 +1,1200 @@
...
+}
+
+float
+PVCamCamera::get_exposure_time_resolution_to_microseconds_() const
+{
+    // get the exposure time resolution
+    int32_t res;
+    if (!is_param_available(PARAM_EXP_RES) ||
+        PV_OK != pl_get_param(hcam_, PARAM_EXP_RES, ATTR_CURRENT, &res)) {
+        // For some older cameras this parameter might not be available
```
Comment (andy-sweet) [inline src/pvcam.cpp:946]: Optional: have you experience this case often? should this case be logged for debugging purposes? As far as I follow the functions and macros, I think it this won't be logged, but I feel like it should.

Optional because I'm a little unsure here and it's not vital on a first pass.

File: src/pvcam.cpp (hunk 14 above)
Comment (andy-sweet) [reply src/pvcam.cpp]: The implementation is simpler and the comment above also clarifies my understanding of the SDK's query to the ROI here, so this is resolved for me.

File: src/pvcam.cpp [hunk 16]
```diff
@@ -0,0 +1,1200 @@
...
+PVCamCamera::maybe_set_output_triggers_(CameraProperties* properties)
+{
+    CHECK(properties);
+}
+
+void
+PVCamCamera::maybe_get_exposure_time_(CameraProperties* properties) const
+{
+    const auto resolution = get_exposure_time_resolution_to_microseconds_();
+    properties->exposure_time_us = (float)exposure_time_ * resolution;
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Why update the properties value from the old/current value of `exposure_time_` here?

File: src/pvcam.cpp [hunk 17]
```diff
@@ -0,0 +1,1200 @@
...
+
+    uint8_t** frame;
+    FRAME_INFO info;
+};
+
+void
+callback_handler(FRAME_INFO* frame_info, void* context)
+{
+    auto* ctx = (CallbackContext*)context;
+    std::unique_lock<std::mutex> frame_lock(*ctx->frame_mutex);
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Nit: since you don't do any manual locking or unlocking, `scoped_lock` is enough here and simpler than `unique_lock`, so I have a mild preference for it.

File: src/pvcam.cpp [hunk 18]
```diff
@@ -0,0 +1,1200 @@
...
+    CHECK(im);
+    CHECK(nbytes);
+    CHECK(info);
+
+    int16_t status;
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    std::unique_lock<std::mutex> frame_lock(*frame_mutex_);
```
Comment (andy-sweet) [inline src/pvcam.cpp:561]: Nit: since you don't do any manual locking or unlocking, scoped_lock is enough here and simpler than unique_lock, so I have a mild preference for it.

File: src/pvcam.cpp [hunk 19]
```diff
@@ -0,0 +1,1200 @@
...
+            .hardware_frame_id = (uint64_t)callback_context_.info.FrameNr,
+        };
+
+    PVCAM(pl_exp_unlock_oldest_frame(hcam_), *pvcam_api_mutex_);
+
+    ++frame_count_;
+
+    rgn_type roi;
+    PVCAM(pl_get_param(hcam_, PARAM_ROI, ATTR_CURRENT, &roi),
+          *pvcam_api_mutex_);
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Why get the ROI here? Does this have some kind of positive side-effect? Or maybe it's a remnant from testing?

File: src/pvcam.cpp [hunk 20]
```diff
@@ -0,0 +1,1200 @@
...
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    std::unique_lock<std::mutex> frame_lock(*frame_mutex_);
+
+    // callback will increment frames_available
+    callback_context_.cv.wait_for(
+      frame_lock,
+      std::chrono::microseconds(exposure_time_ * exposure_time_to_us_),
```
Comment (andy-sweet) [inline src/pvcam.cpp]: I guess the reason we wait for up to the *exact frame time* here is because Acquire is constantly polling for frames at the equivalent frame rate, so this roughly prevents Acquire's source/camera thread from backing up - is that right?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Looks great now! Pretty much ready to approve. Just have a few remaining questions, some of which are around the exposure time state of `PVCamCamera`.

File: src/pvcam.cpp (hunk 15 above)
Comment (aliddell) [reply src/pvcam.cpp:946]: I don't think it should be logged here. From the documentation, it seemed to me that Photometrics originally did not expose this value as a parameter and the exposure time was always a millisecond. We also don't expose this parameter and it's purely for internal calculation.

File: src/pvcam.cpp (hunk 18 above)
Comment (aliddell) [reply src/pvcam.cpp:561]: Normally I prefer `scoped_lock` as well, but the `callback_context_.cv.wait_for(...)` a few lines down doesn't accept a `scoped_lock`, probably because `scoped_lock` doesn't have an `unlock()` method.

File: src/pvcam.cpp (hunk 20 above)
Comment (aliddell) [reply src/pvcam.cpp]: That's right, or that was the idea anyway.

File: src/pvcam.cpp (hunk 19 above)
Comment (aliddell) [reply src/pvcam.cpp]: Yeah, this looks like a holdover from testing. Thanks for catching.

File: src/pvcam.cpp (hunk 16 above)
Comment (aliddell) [reply src/pvcam.cpp]: I can't remember exactly why I did this. All I do remember is that there was a good reason to cache the exposure time alongside the image ROI and pixel type (see comment at the bottom of the file), but looking at this it doesn't make sense. If the exposure time parameter isn't available in line 683, it's not ever going to be available and the exposure time will always be 0. Might have been an artifact of testing. I'll pull it out.

File: src/pvcam.cpp (hunk 15 above)
Comment (andy-sweet) [reply src/pvcam.cpp:946]: OK, that's good enough justification for me. Resolved.

File: src/pvcam.cpp (hunk 20 above)
Comment (andy-sweet) [reply src/pvcam.cpp]: I guess the timed wait also guards against the callback throwing, so I think I'm OK with this.

However, I feel like this is an important design decision, so it could be worth reviewing this in person and/or getting feedback from @nclack  here.

**Review by andy-sweet — APPROVED** [review]
Looks good to me now, but I feel like this one might be worth a live review with @nclack - even just a short one at standup.

File: src/pvcam.cpp [hunk 21]
```diff
@@ -550,16 +560,33 @@ PVCamCamera::get_frame(void* im, size_t* nbytes, struct ImageInfo* info)
...
+        if (callback_context_.error) {
+            // error is logged in the callback handler thread
+            callback_context_.error = false;
+            *nbytes = 0;
+            return;
+        }
+
+        if (callback_context_.frames_available > 0) {
+            break;
+        }
```
Comment (andy-sweet) [inline src/pvcam.cpp]: Nit: couldn't this condition be absorbed into the while's condition?

I.e. use `while (is_running && callback_context_.frames_available == 0)` and remove these three lines?

File: src/pvcam.cpp [hunk 22]
```diff
@@ -0,0 +1,1218 @@
...
+{
+    CHECK(im);
+    CHECK(nbytes);
+    CHECK(info);
+
+    std::unique_lock<std::mutex> frame_lock(*frame_mutex_);
+
+    while (is_running_) {
+        // callback will increment frames_available
+        callback_context_.cv.wait_for(
```
Comment (andy-sweet) [inline src/pvcam.cpp:565]: With the error handling added, I thought that `wait` without the while loop might be enough here.

But I didn't think about the case that `stop` is called while `get_frame` is waiting. In that case, I guess the callback is not executed or none of the variables we're waiting on change appropriately? Is that the case this `while` loop plus `wait_for` is handling?

**Review by andy-sweet — APPROVED** [review]
I'm still a little confused by `get_frame`. But I'm happy enough to get this in, especially as things may change a little with triggering.

File: src/pvcam.cpp (hunk 21 above)
Comment (aliddell) [reply src/pvcam.cpp]: Sure, that's a good idea.

File: src/pvcam.cpp (hunk 22 above)
Comment (aliddell) [reply src/pvcam.cpp:565]: Yeah, I found out to my disappointment that you could call `camera_stop()` in the main thread ~~or the sink thread~~, while the source thread would hang on `camera_get_frame()` forever.
